### PR TITLE
Make FindReflectionOnline more configurable

### PR DESCRIPTION
**Pull request by moufmouf**
This PR closes #285 .

It adds to the `FindReflectionOnline` constructor a `SourceLocator` parameter making it more configurable.

Now, `FindReflectionOnline` can be used on classes that extend/implement some classes/interfaces that need autoloading.

In order to prevent BC breaks, the `$sourceLocator` parameter is nullable. This is not ideal but it keeps compatibility. I recommend removing the "nullable" part on the next major version.

Also, I added a utility `buildDefaultFinder` static method that mimics the `ClassReflector::buildDefaultReflector` method.

I hope you'll find this PR useful! Don't hesitate if you have any comments.

### CODE REVIEW

File: src/Util/FindReflectionOnLine.php [hunk 1]
```diff
@@ -58,7 +85,11 @@ public function __invoke(string $filename, int $lineNumber)
     private function computeReflections(string $filename) : array
     {
         $sourceLocator = new SingleFileSourceLocator($filename);
-        $reflector = new ClassReflector($sourceLocator);
+        if ($this->sourceLocator !== null) {
+            $reflector = new ClassReflector(new AggregateSourceLocator([$this->sourceLocator, $sourceLocator]));
```
Comment (Ocramius) [inline src/Util/FindReflectionOnLine.php]: Simplify this using `??` instead of an if/else

File: src/Util/FindReflectionOnLine.php [hunk 2]
```diff
@@ -8,11 +8,38 @@
...
+use Roave\BetterReflection\SourceLocator\Type\SourceLocator;
 
 final class FindReflectionOnLine
 {
     /**
+     * @var SourceLocator
+     */
+    private $sourceLocator;
+
+    public function __construct(SourceLocator $sourceLocator = null)
```
Comment (Ocramius) [inline src/Util/FindReflectionOnLine.php]: Changing the parameter for it to be required would be better

Comment (Ocramius) [inline src/Util/FindReflectionOnLine.php]: Yes, it would be a documented BC break

**Comment by asgrim** [comment]
Hmm, not sure if this is gonna work; I see the problem, but the issue with this solution is we will reflect everything in a source locator given, regardless of which file it's in I think?

**Comment by moufmouf** [comment]
@Ocramius : I made the constructor argument compulsory, as per your comment (and your defensive best practices :) ) Is there anywhere I can document this BC breaking change?

@asgrim Not sure I understand your comment. The code is basically [aggregating the source locator you pass in parameter with its own `SingleFileSourceLocator`](https://github.com/moufmouf/BetterReflection/blob/d4f379037c63e59f68f1d9ebb0774dcc2d5fc623/src/Util/FindReflectionOnLine.php#L88).  `SingleFileSourceLocator` is used to locate the class/methods in the file, and the sourceLocator you pass in parameter of the constructor can be used to load dependencies (extended/implemented classes/interfaces). Is there something that could go wrong here?

**Comment by asgrim** [comment]
@moufmouf on further inspection, my worries have been settled. I hadn't read through the code thoroughly enough and was worried that the aggregated locator (i.e. along with anything else, potentially a lot of code) was being used and having the `locateIdentifiersByType` called on it, but it's ok, it's still using the `$singleFileSourceLocator->locateIdentifiersByType(..)` so shouldn't reflect everything (that would be potentially very slow).

**Comment by moufmouf** [comment]
@asgrim Done!

**Pull request merged**